shared/utility/motion/InverseKinematics.hpp

```diff
@@ -52,21 +52,167 @@ namespace utility::motion::kinematics {
+        hipRoll = (hipRollPositive ? 1 : -1) * std::acos(std::max(std::min(cosHipRoll, Scalar(1)), Scalar(-1)));
+
+        const Scalar phi4 = M_PI - knee - lowerLeg;
+        // Superposition values:
+        const Scalar sinPIminusPhi2 = std::sin(M_PI - phi2);
+        const Eigen::Matrix<Scalar, 3, 1> unitUpperLeg =
+            unitTargetLeg * (std::sin(phi2 - phi4) / sinPIminusPhi2) + ankleY * (std::sin(phi4) / sinPIminusPhi2);
+        const bool isHipPitchPositive = hipX.dot(unitUpperLeg.cross(legPlaneGlobalZ)) >= 0;
+
+        hipPitch = (isHipPitchPositive ? 1 : -1)
+                   * std::acos(std::max(std::min(legPlaneGlobalZ.dot(unitUpperLeg), Scalar(1)), Scalar(-1)));
+
+        // If leg is above waist then hipX is pointing in the wrong direction in the xy plane
+        Eigen::Matrix<Scalar, 3, 1> hipXProjected = (isAnkleAboveWaist ? -1 : 1) * hipX;
+        hipXProjected.z()                         = 0;
+        hipXProjected.normalize();
+        const bool isHipYawPositive = hipXProjected.dot(Eigen::Matrix<Scalar, 3, 1>::UnitY()) >= 0;
+        hipYaw                      = (isHipYawPositive ? 1 : -1)
+                 * std::acos(std::max(std::min(hipXProjected.dot(Eigen::Matrix<Scalar, 3, 1>::UnitX()), Scalar(1)),
+                                      Scalar(-1)));
+
+        if (limb == LimbID::LEFT_LEG) {
+            positions.push_back(std::make_pair(ServoID::L_HIP_YAW, -hipYaw));
+            positions.push_back(std::make_pair(ServoID::L_HIP_ROLL, hipRoll));
+            positions.push_back(std::make_pair(ServoID::L_HIP_PITCH, -hipPitch));
+            positions.push_back(std::make_pair(ServoID::L_KNEE, M_PI - knee));
+            positions.push_back(std::make_pair(ServoID::L_ANKLE_PITCH, -anklePitch));
+            positions.push_back(std::make_pair(ServoID::L_ANKLE_ROLL, ankleRoll));
+        }
+        else {
+            positions.push_back(std::make_pair(ServoID::R_HIP_YAW, (model.leg.LEFT_TO_RIGHT_HIP_YAW) * -hipYaw));
+            positions.push_back(std::make_pair(ServoID::R_HIP_ROLL, (model.leg.LEFT_TO_RIGHT_HIP_ROLL) * hipRoll));
+            positions.push_back(std::make_pair(ServoID::R_HIP_PITCH, (model.leg.LEFT_TO_RIGHT_HIP_PITCH) * -hipPitch));
+            positions.push_back(std::make_pair(ServoID::R_KNEE, (model.leg.LEFT_TO_RIGHT_KNEE) * (M_PI - knee)));
+            positions.push_back(
+                std::make_pair(ServoID::R_ANKLE_PITCH, (model.leg.LEFT_TO_RIGHT_ANKLE_PITCH) * -anklePitch));
+            positions.push_back(
+                std::make_pair(ServoID::R_ANKLE_ROLL, (model.leg.LEFT_TO_RIGHT_ANKLE_ROLL) * ankleRoll));
+        }
+
+        return positions;
+    }
+
+    template <typename Scalar>
+    [[nodiscard]] std::vector<std::pair<ServoID, Scalar>> calculateLegJoints(
         const message::motion::KinematicsModel& model,
         const Eigen::Affine3f& leftTarget,
-        const Eigen::Affine3f& rightTarget);
+        const Eigen::Affine3f& rightTarget) {
```

Comment:
Same for the `leftTarget`
```suggestion
        const Eigen::Transform<Scalar, 3, Eigen::Affine>& rightTarget) {
```